src/sif/bicyclecost.cc

```diff
@@ -592,6 +586,20 @@ Cost BicycleCost::EdgeCost(const baldr::DirectedEdge* edge,
                               kGradeBasedSpeedFactor[edge->weighted_grade()]) +
                              0.5f);
 
+  if (shortest_) {
+    // Only give it the modulated bike speed if edge is no ferry route or has steps
+    speed = edge->use() == Use::kSteps ? 1 : (edge->use() == Use::kFerry ? speed : bike_speed);
+    assert(speed < speedfactor_.size());
+    float sec = (edge->length() * speedfactor_[speed]);
+    return Cost(edge->length(), sec);
+  }
+
+  // Stairs/steps - high cost (travel speed = 1kph) so they are generally avoided.
+  if (edge->use() == Use::kSteps) {
+    float sec = (edge->length() * speedfactor_[1]);
+    return {sec * kBicycleStepsFactor, sec};
+  }
+
```

Comment:
rearranged a little so `bike_speed` won't be uselessly computed if there's a ferry. Also kept all duration-relevant conditions, was a serious oversight before on my part..

Reply:
unfortunately there are still some missed spots. if its a ferry its going to bail early and ignore shortest. also at the very end of the function (ie the normal case doesnt consider shortest). remember that for shortest we want to retain the original time estimate (`Cost::sec`) but we just want to set the cost part (`Cost::cost`) to edge length. i think the easiest way to understand that change would be a ternary in every return. i'll re-organize it and you can have a look to see what i mean

Reply:
:facepalm: